subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+import org.apache.commons.lang.WordUtils;
+
+import java.util.Optional;
+
+/**
+ * Defines {@link ConfigurationRole}s representing common allowed usage patterns.
+ *
+ * These should be preferred over defining custom roles; whenever possible.  Use {@link #byUsage(boolean, boolean, boolean, boolean, boolean, boolean)}
+ * to attempt to locate a matching role by its usage characteristics.
+ *
+ * @since 8.0
+ */
+public enum ConfigurationRoles implements ConfigurationRole {
+    /**
+     * An unrestricted configuration, which can be used for any purpose.
+     *
+     * This is available for backwards compatibility, but should not be used for new configurations.  It is
```

Comment:
I think its premature to deprecate a legacy configuration altogether without first designing a solution for the very common resolvable/bucket use-case. I thought we discussed in Denver making the legacy configuration a resolvable/bucket that was only deprecated for consumption.

Reply:
Right now, what we've been calling LEGACY is the maximally permissible configuration, but in some way it's also the role of a configuration which is created when not using one of the factory methods (just calling create or register). Maybe we can choose a different name for it if you want to leave the legacy terminology as meaning the maximally permissible configuration, though I think we will need a role name for resolvable/bucket.

Reply:
I'm fine with adding another role if we want it.  Maybe `DECLARING_RESOLVABLE`?

Reply:
I'm going to go with `INTENDED_RESOLVABLE_BUCKET`, and `@Deprecated` it immediately, just like `LEGACY`.